src/NHibernate.Test/NHSpecificTest/GH0831/FixtureByCode.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Mapping.ByCode;
+
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH0831
+{
+	public class ByCodeFixture : TestCaseMappingByCode
+	{
+		private readonly IList<Entity> entities = new List<Entity>
+		{
+			new Entity { EntityValue = 0.5m },
+			new Entity { EntityValue = 1.0m },
+			new Entity { EntityValue = 1.5m },
+			new Entity { EntityValue = 2.0m },
+			new Entity { EntityValue = 2.5m },
+			new Entity { EntityValue = 3.0m }
+		};
+
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<Entity>(rc =>
+			{
+				rc.Id(x => x.Id, m => m.Generator(Generators.GuidComb));
+				rc.Property(x => x.EntityValue);
+			});
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override void OnSetUp()
+		{
+			using (ISession session = OpenSession())
+			using (ITransaction transaction = session.BeginTransaction())
+			{
+				foreach (Entity entity in entities)
+				{
+					session.Save(entity);
+				}
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (ISession session = OpenSession())
+			using (ITransaction transaction = session.BeginTransaction())
+			{
+				session.Delete("from System.Object");
+
+				session.Flush();
+				transaction.Commit();
+			}
+		}
+
+		[Test]
+		public void CanHandleAdd()
+		{
+			CanHandle(e => decimal.Add(e.EntityValue, 2) > 3.0m);
```

Comment:
Its failure on Firebird seems to me due to some Firebird limitation. It looks like the operation on the column is enough to cause Firebird to lose track of the data type.
Maybe a workaround could be to add a cast, but I think we should just ignore that Firebird limitation with something like:
```c#
Assume.That(TestDialect.SupportsNonDataBoundCondition, Is.True);
```

Granted, here the condition is data bound, but the addition (or subtraction) on the data causes the same trouble. And I lack an idea for a new `TestDialect` property name for such a case.